### ISSUE TITLE
Some Filter not produce stream

### ISSUE THREAD

**Issue by lucemia** [livingbio/typed-ffmpeg#22]
e.g. 
vif filter https://ffmpeg.org/ffmpeg-filters.html#vif

<details>
  <summary><a href="https://ffmpeg.org/ffmpeg-filters.html#vif">https://ffmpeg.org/ffmpeg-filters.html#vif</a></summary>

- FFmpeg Filters Documentation
- 1 Description 2 Filtering Introduction 4 Filtergraph description. 5 Timeline editing 6 Changing options at runtime with a command 7 Options for filters with several inputs (framesync) 8 Audio Filters. 8.5.1 Examples.
</details>

**Comment by lucemia** [comment]
The `vif` filter in FFmpeg, which stands for "Video Information Filter," is a bit different from conventional filters as it's primarily used for analyzing video streams rather than processing or altering them. This filter calculates and outputs video quality metrics, which can be used for various purposes like quality assessment or debugging.

When using filters like `vif` in `-filter_complex`, it's important to understand that they don't output a traditional video or audio stream that can be directly used for encoding. Instead, they output information, often in the form of text or data values. Here's how you can use such filters in a `-filter_complex` chain:

1. **Analysis in Parallel**: You can use the `vif` filter alongside other processing filters to analyze the video stream while also performing other operations. For example, you might be scaling a video and simultaneously wanting to analyze its quality before and after scaling.

2. **Output to File or Console**: The results from the `vif` filter can be directed to a file or printed to the console. This is typically done using the `metadata` option in FFmpeg, which allows you to export the analysis data.

3. **Filtergraph Example**: Here's a basic example of how you might set up a filtergraph with the `vif` filter:

   ```
   ffmpeg -i input.mp4 -filter_complex "[0:v]split=2[main][vif]; [vif]vif=psnr=1:ssim=1[analysis]" -map "[main]" output.mp4
   ```

   In this command, the `split` filter is used to create two copies of the video stream: one for processing/output (`[main]`) and one for analysis (`[vif]`). The `vif` filter is then applied to the `[vif]` stream to analyze it, and the results can be logged or printed.

4. **Reviewing the Output**: The analysis results from the `vif` filter will be available in the FFmpeg log. Depending on your FFmpeg configuration, you might need to adjust the log level to ensure that the results are printed to the console or to a log file.

In summary, filters like `vif` that do not output a traditional stream are still usable in `-filter_complex`. They are typically used for analysis purposes, and their output is in the form of data or metrics rather than a video or audio stream. The setup in `-filter_complex` involves routing the video stream through the filter and capturing the analysis output, often through logging.

**Comment by lucemia** [comment]
The code snippet you've provided is a part of the FFmpeg filter registration and it does provide some insights into what the filter does and how it behaves. To determine whether a filter outputs a stream or not, you can look at certain elements in this structure.

In your example:

1. **`FILTER_INPUTS(vif_inputs)` and `FILTER_OUTPUTS(vif_outputs)`**: These macros define the inputs and outputs of the filter. If a filter has output streams, they will be defined here. For filters like `vif` that are primarily used for analysis, the output might not be a traditional video or audio stream but could be metadata or logs.

2. **`.flags` Field**: The `.flags` field can give you additional context. For instance, in the `vif` filter, it includes `AVFILTER_FLAG_METADATA_ONLY`. This flag suggests that the filter is used for generating metadata rather than outputting a modified video or audio stream. This is a strong indicator that the filter's primary purpose is analysis or information extraction rather than stream processing.

3. **Filter Functionality**: The `.name` and `.description` fields give you an idea of what the filter does. In your example, the `vif` filter calculates the VIF (Video Information Filter) between two video streams, which implies it's used for analysis.

4. **Implementation Details**: To get a definitive answer, you might have to look into the implementation details of the filter, particularly the `activate` function or other custom processing functions. This would show you exactly how the filter processes input and what kind of output it generates.

In summary, while the structure you provided gives some clues (especially the `AVFILTER_FLAG_METADATA_ONLY` flag), it may not always be sufficient to definitively determine the output type of a filter. For a complete understanding, examining the implementation of the filter's processing functions would be necessary. In the case of `vif`, the presence of `AVFILTER_FLAG_METADATA_ONLY` suggests that it's more about metadata generation (like producing analysis data) rather than creating a modified video/audio stream.
